javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll

```diff
@@ -108,3 +117,317 @@ private class NosqlInjectionSinkCharacteristic extends EndpointCharacteristic {
     confidence = 1.0
   }
 }
+
+/*
+ * Characteristics that are indicative of not being a sink of any type.
+ */
+
+/**
+ * A characteristic that is an indicator of not being a sink of any type, because it's an argument that has a manual
+ * model.
+ */
+abstract private class OtherModeledArgumentCharacteristic extends EndpointCharacteristic {
+  bindingset[this]
+  OtherModeledArgumentCharacteristic() { any() }
+}
+
+/**
+ * A characteristic that is an indicator of not being a sink of any type, because it's an argument to a function of a
+ * builtin object.
+ */
+abstract private class ArgumentToBuiltinFunctionCharacteristic extends OtherModeledArgumentCharacteristic {
```

Comment:
These are some graphs I made for my own comprehension, but they may be useful for others.

## Main

This is the hierarchy we have on `main` (but with the new `OtherModeledArgumentCharacteristic` added in):

```mermaid
graph TD;
    string-->EndpointC;
    EndpointC-->DomBasedXssSinkC;
    EndpointC-->TaintedPathSinkC;
    EndpointC-->SqlInjectionSinkC;
    EndpointC-->NosqlInjectionSinkC;
    EndpointC-->OtherModeledArgumentC*;
```

## PR
This is the to-be-added sub-hierarchy. You'll have to zoom in, or read the markdown of this comment, or own a huge screen :)

In addition to what I said above,
* I think it is probably wrong that `LikelyNotASinkCharacteristic` and `NotASinkCharacteristic` are subclasses of `OtherModeledArgumentCharacteristic`. For example, there should be the possibility for building not-a-sinks that are NOT modeled arguments.
* I don't think that all builtin functions are also modeled, therefore `ArgumentToBuiltinFunctionCharacteristic` should not extend `OtherModeledArgument`.
* 

```mermaid
graph TD;
    OtherModeledArgumentC*-->ArgToBuiltinFuncC*;
    OtherModeledArgumentC*-->NotASinkC*;
    OtherModeledArgumentC*-->LikelyNotASinkC*;
    ArgToBuiltinFuncC*-->ArgToArrayC*;
    ArgToBuiltinFuncC*-->ArgToBuiltinGlobalVarRefC*;
    ArgToBuiltinFuncC*-->ConstantReceiverC*;
    ArgToBuiltinFuncC*-->BuiltinCallNameC*;
    NotASinkC*-->Lodash*;
    NotASinkC*-->JQArgC*;
    NotASinkC*-->ClientRequestC*;
    NotASinkC*-->PromiseDefC*;
    NotASinkC*-->CryptoKeyC*;
    NotASinkC*-->LoggerMethodC*;
    NotASinkC*-->TimeoutC*;
    NotASinkC*-->ReceiverStorageC*;
    NotASinkC*-->StringStartsWithC*;
    NotASinkC*-->StringEndsWithC*;
    NotASinkC*-->StringRegExpTestC*;
    NotASinkC*-->EvRegistrationC*;
    NotASinkC*-->EvDispatchC*;
    NotASinkC*-->MemberCandTC*;
    NotASinkC*-->FSAccessC*;
    NotASinkC*-->DBAccessC*;
    NotASinkC*-->DomC*;
    NotASinkC*-->NextFuncCallC*;
    NotASinkC*-->DojoRequireC*;
    NotASinkC*-->B64ManipC*;
    LikelyNotASinkC*-->ArgToArrayC*;
    LikelyNotASinkC*-->ArgToBuiltinGlobalVarRefC*;
```

Reply:
AIUI, the hierarchy in the PR directly reproduces the hierarchy on main. Is that untrue? It's entirely possible I messed something up, but that was my intention.

As for improving on the hierarchy that currently exists on main, I'm 💯 happy to do so. In this PR I simply attempted to reproduce main with the new framework, because I don't have the JS knowledge to make assertions such as "not all builtin functions are also modeled". You _do_ have that knowledge, so I think you're the right person to make improvements to the hierarchy. I would suggest doing that in a subsequent PR, that you author and someone else with JS knowledge reviews, so as to separate the switch to the new framework from substantive changes to the way we categorize endpoints.

Reply:
> I think that the whole class-hierarchy starting at `OtherModeledArgumentCharacteristic` could be merged into one `private class OtherModeledArgumentCharacteristic` that has a `getImplications` implementation that contains the disjunction of all the classes' `getImplications` that you're proposing to add here. If not that, at least some classes in the sub-hierarchy could be merged in this fashion.
> 
> Or is there a specific reason to use such a big class hierarchy?
> 
> Any opinion, @henrymercer?

Did you mean `getEndpoints` rather than `getImplications`?

Potentially we could get rid of `OtherModeledArgumentCharacteristic` and `ArgumentToBuiltinFunctionCharacteristic` entirely, because at the moment they don't implement any predicates, and all they do is preserve the hierarchy that existed in the old code, with no actual implications (that is, if we removed them, it would change nothing about the current selection of endpoints). Again I would prefer to do this in a subsequent PR that simplifies/improves the existing hierarchy, rather than in this one.

We could also possibly get rid of all the classes that inherit from `NotASinkCharacteristic`, and implement them as a single disjunction of all the classes' `getEndpoints`, as you suggested, pairing one string description with each selection of endpoints. We could then do the same for all the classes that inherit from `LikelyNotASinkCharacteristic`. That would make the code shorter, but IMHO it would make it less clear. Right now we have one class per characteristic, which translates directly into one string in the list of characteristics we'll assign to each training endpoint. If a couple of these are really similar we can merge them together, but overall I think we're better off with one `EndpointCharacteristic` class per logical characteristic. This makes it very easy to experiment with changes to one characteristic without accidentally affecting anything else, and is a fairly critical advantage of the new design.

Incidentally, the list of characteristics _does_ get used by the training code. Suppose I have 1000 training samples with `ClientRequest` but only 10 samples with `PromiseDefinition`. If we train the model naively, to minimize the loss it will try to learn the patterns behind `ClientRequest`, but won't care too much whether or not it learns the patterns behind `PromiseDefinition`. To prevent this, I give higher weight in the loss function to samples that have `PromiseDefinition` in their list of characteristics. This pushes the model to learn these less common patterns as well as the most common ones. When I first implemented this change in the modeling code it had significant impact, though I haven't experimented with turning it off any time lately, so I don't know how big the impact is these days.

In the future I think we might also want to subsample negative samples that have very common characteristics, so that we can achieve a smaller training set without sacrificing the diversity of our negative samples.

Reply:
From a quick look, this PR introduces a new class hierarchy based on the implementation of the predicates in `CoreKnowledge`.  This would concern me a little in terms of BDD usage, and I don't think it's the right long term solution, as Stephan points out.

That said, if performance is OK, then I don't have quealms with introducing a temporary complex class hierarchy, as long as we remove it by the end of this refactoring.

Reply:
I'm happy to push a commit, but can you tell me exactly what hierarchy you want? Incidentally, when I spoke of a subsequent PR, I meant immediately following this one, as soon as this one is merged. It's a trivial change to make, though, as long as I'm clear on your desired hierarchy. If it's easier than describing it in English, you can even create a branch off of this one to implement the change you want there. Rather than merging it into this PR, I'll copy it manually so you're not a committer 😊. If possible, I'd love to get this PR merged today, before my long weekend.

Reply:
After our conversation yesterday, I'm convinced this won't get lost or forgotten. It's fine to do it in a later PR! 👍